### ISSUE TITLE
Refactor `extra.numpy.basic_indices()` handling of `max_dims` when `allow_newaxis=False`

### ISSUE THREAD

**Issue by Zac-HD** [HypothesisWorks/hypothesis#3091]
Via https://github.com/HypothesisWorks/hypothesis/pull/3065#issuecomment-916756585, `basic_indices()` is doing more work than it needs to when `allow_newaxis=False` - and that's particularly salient because the `array_api` extra doesn't have `newaxis` at all.  So the internals part:

- [ ] When `allow_newaxis=False`, `max_dims` should be set to `len(shape)`, and
- [ ] We can therefore move it to `_array_helpers.py` and refactor the `array_api` extra to use this too

This also suggests a constraint on the arguments that we're not currently enforcing for the Numpy extra:

```python
if max_dims > len(shape) and not allow_newaxis:
    note_deprecation(..., since="RELEASEDAY", has_codemod=False)
    # Becomes `raise InvalidArgument(...)` in the next major version
```

**Comment by honno** [comment]
Happy to do this—will make a PR next week.

For reference this was the deprecation notice I had written before. I *think* it covers the right things, I just need to iterate.
```python
>>> nps.basic_indices((3, 3, 3), max_dims=5).example()
HypothesisDeprecationWarning: max_dims=5 and allow_newaxis=False, so we cannot generate indices that would produce arrays with dimensions greater than len(shape)=3 i.e. the dimensions of the indexed array. In the next release, max_dims would default to the the dimensions of shape=(3, 3, 3) when allow_newaxis=False, and passing max_dims greater than len(shape) will raise an error.
```

**Comment by Zac-HD** [comment]
Closed by #3098.
